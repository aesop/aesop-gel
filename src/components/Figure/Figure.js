import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import styles from './Figure.module.css';

const Figure = ({
  caption,
  children,
  className,
  hasCaptionBorder,
  hasCaptionOffset,
  id,
  theme,
}) => {
  const classSet = cx(styles.base, styles[theme], className);

  const captionWrapperClassSet = cx(styles.captionWrapper, {
    [styles.border]: hasCaptionBorder && caption,
    [styles.captionOffset]: hasCaptionOffset && caption,
  });

  return (
    <figure className={classSet} id={id}>
      {children}
      {caption && (
        <figcaption className={captionWrapperClassSet}>
          <div className={styles.caption}>{caption}</div>
        </figcaption>
      )}
    </figure>
  );
};

Figure.propTypes = {
  caption: PropTypes.any,
  children: PropTypes.node,
  className: PropTypes.string,
  hasCaptionBorder: PropTypes.bool,
  hasCaptionOffset: PropTypes.bool,
  id: PropTypes.string,
  theme: PropTypes.oneOf(['dark', 'light']),
};

Figure.defaultProps = {
  caption: undefined,
  children: undefined,
  className: undefined,
  hasCaptionBorder: true,
  hasCaptionOffset: true,
  id: undefined,
  theme: 'dark',
};

export { Figure };
