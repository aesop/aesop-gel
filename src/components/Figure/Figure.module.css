@import 'styles/index';

.base {
  margin: 0;

  &,
  & * {
    box-sizing: border-box;
  }
}

.caption {
  padding-top: 25px;
  padding-right: 20px;
  padding-bottom: 25px;
  padding-left: 20px;
  margin: 0;
  font-family: var(--font-family-sans-serif-medium);
  font-size: rem(14px);
  line-height: 1.7;

  @media (--viewport-md) {
    padding-right: 25px;
    padding-left: 34px;
  }

  @media (--viewport-lg) {
    padding-left: 20px;
  }

  .dark & {
    color: var(--color-dark-label);
  }

  .light & {
    color: var(--color-light-label);
  }

  p {
    &:first-of-type {
      margin-top: 0;
    }

    &:last-of-type {
      margin-bottom: 0;
    }
  }
}

.captionWrapper {
  margin: 0;
}

.captionOffset {
  @media (--viewport-md) {
    margin-left: 27.34%;
  }

  @media (--viewport-lg) {
    margin-left: 40.39%;
  }
}

.border {
  border-bottom: 1px solid var(--color-grey-rgba-20);
}
