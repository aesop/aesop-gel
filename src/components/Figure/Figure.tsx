import React, { FunctionComponent } from 'react';
import cx from 'classnames';
import styles from './Figure.module.css';

interface FigureProps {
  caption?: string;
  children?: React.ReactNode;
  className?: string;
  hasCaptionBorder?: boolean;
  hasCaptionOffset?: boolean;
  id?: string;
  theme?: string;
  foo?: boolean;
}

const Figure: FunctionComponent<FigureProps> = ({
  caption,
  children,
  className,
  hasCaptionBorder,
  hasCaptionOffset,
  id,
  theme = 'dark',
}) => {
  const classSet = cx(styles.base, styles[theme], className);

  const captionWrapperClassSet = cx(styles.captionWrapper, {
    [styles.border]: hasCaptionBorder && caption,
    [styles.captionOffset]: hasCaptionOffset && caption,
  });

  return (
    <figure className={classSet} id={id}>
      {children}
      {caption && (
        <figcaption className={captionWrapperClassSet}>
          <div className={styles.caption}>{caption}</div>
        </figcaption>
      )}
    </figure>
  );
};

export { Figure };
