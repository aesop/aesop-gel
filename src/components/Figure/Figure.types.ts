type Props = {
  caption?: string;
  children?: React.ReactNode;
  className?: string;
  hasCaptionBorder?: boolean;
  hasCaptionOffset?: boolean;
  id?: string;
  theme?: string;
  foo?: boolean;
};

export type { Props };
