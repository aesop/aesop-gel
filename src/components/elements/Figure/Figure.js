import React from 'react';
import cx from 'classnames';
import parse from 'html-react-parser';
import marked from 'marked';
import Props from '~/types/Figure.types';
import styles from './Figure.module.css';

const Figure: React.FunctionComponent<Props> = ({
  caption,
  children,
  classSet,
  id,
}) => {
  const className: string = cx(styles.base, classSet);

  return (
    <figure className={className} id={id}>
      {children}
      {caption && (
        <figcaption className={styles.captionWrapper}>
          <div className={styles.caption}>{parse(marked(caption))}</div>
        </figcaption>
      )}
    </figure>
  );
};

export default Figure;
