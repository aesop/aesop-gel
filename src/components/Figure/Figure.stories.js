import React from 'react';
import { storiesOf } from '@storybook/react';
import { text } from '@storybook/addon-knobs';
import 'normalize.css';

import '~/styles/base.module.css';
import '~.storybook/storybook.module.css';
import Figure from './Figure';
import Image from '~/components/Image';

storiesOf('Figure', module).add('Base component', () => (
  <Figure caption={text('caption', 'About Aesop')} id={text('id', 'aesop')}>
    <Image
      altText={text('altText', 'Image of Milk Cleansing Campaign Aesop 2019')}
      large={text(
        'large',
        'https://www.aesop.com/medias/GelCleanser-640x640.png?context=bWFzdGVyfGltYWdlc3w1MzE5NTl8aW1hZ2UvcG5nfGltYWdlcy9oOTUvaGNkLzg4MDc3OTc4NTAxNDIucG5nfGM3NTNhYjdkMWNiOTRhNWZhNDA0NWExZWJiODMxN2VhNzEyY2ZkNGUyYzc4ZWVkZDRjYTIzMzZmZTRlOGYxNjM',
      )}
    />
  </Figure>
));
