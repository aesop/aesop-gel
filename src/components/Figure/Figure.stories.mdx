import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { withKnobs } from '@storybook/addon-knobs';
import { text, select } from '@storybook/addon-knobs';
import Figure from './Figure';
import Image from '~/components/Image';

<Meta title="Components/Figure" component={Figure} decorators={[withKnobs]} />

# Figure

<Preview>
  <Story name="Base component">
    <Figure
      caption={text('caption', 'About Aesop')} id={text('id', 'aesop')}
      theme={select('Image theme', ['dark', 'light'], 'dark')}
    >
      <Image
        altText={text('altText', 'Image of Milk Cleansing Campaign Aesop 2019')}
        large={text(
          'large',
          'https://www.aesop.com/medias/GelCleanser-640x640.png?context=bWFzdGVyfGltYWdlc3w1MzE5NTl8aW1hZ2UvcG5nfGltYWdlcy9oOTUvaGNkLzg4MDc3OTc4NTAxNDIucG5nfGM3NTNhYjdkMWNiOTRhNWZhNDA0NWExZWJiODMxN2VhNzEyY2ZkNGUyYzc4ZWVkZDRjYTIzMzZmZTRlOGYxNjM',
        )}
        medium={text(
          'medium',
          'https://www.aesop.com/medias/GelCleanser-640x640.png?context=bWFzdGVyfGltYWdlc3w1MzE5NTl8aW1hZ2UvcG5nfGltYWdlcy9oOTUvaGNkLzg4MDc3OTc4NTAxNDIucG5nfGM3NTNhYjdkMWNiOTRhNWZhNDA0NWExZWJiODMxN2VhNzEyY2ZkNGUyYzc4ZWVkZDRjYTIzMzZmZTRlOGYxNjM',
        )}
        small={text(
          'small',
          'https://www.aesop.com/medias/GelCleanser-640x640.png?context=bWFzdGVyfGltYWdlc3w1MzE5NTl8aW1hZ2UvcG5nfGltYWdlcy9oOTUvaGNkLzg4MDc3OTc4NTAxNDIucG5nfGM3NTNhYjdkMWNiOTRhNWZhNDA0NWExZWJiODMxN2VhNzEyY2ZkNGUyYzc4ZWVkZDRjYTIzMzZmZTRlOGYxNjM',
        )}
      />
    </Figure>
  </Story>
</Preview>

## API

<Props of={Figure} />

## Usage Example

```
<Figure caption="About Aesop">
  <Image
    altText="Image of Milk Cleansing Campaign Aesop 2019"
    large="https://www.aesop.com/medias/GelCleanser-640x640.png?context=bWFzdGVyfGltYWdlc3w1MzE5NTl8aW1hZ2UvcG5nfGltYWdlcy9oOTUvaGNkLzg4MDc3OTc4NTAxNDIucG5nfGM3NTNhYjdkMWNiOTRhNWZhNDA0NWExZWJiODMxN2VhNzEyY2ZkNGUyYzc4ZWVkZDRjYTIzMzZmZTRlOGYxNjM"
  />
</Figure>

```
