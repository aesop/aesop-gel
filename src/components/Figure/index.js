/* istanbul ignore file */
import { Figure } from './Figure';

export { Figure };
