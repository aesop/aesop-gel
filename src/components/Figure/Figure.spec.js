import React from 'react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import Figure from './Figure';
import Image from '~/components/Image';

configure({ adapter: new Adapter() });

describe('<Figure />', () => {
  it('should be defined', () => {
    expect(Figure).toBeDefined();
  });

  it('renders base component correctly', () => {
    const tree = renderer
      .create(
        <Figure caption="About Aesop" id="about">
          <Image
            altText="Image Alt Text"
            large="https://release.aesop-web-ui.aesopdigital.io/images/products/Skin_Perfect_Facial_Hydrating_Cream_60mL_large.png"
          />
        </Figure>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
