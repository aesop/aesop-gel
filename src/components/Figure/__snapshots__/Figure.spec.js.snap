// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Figure /> renders base component correctly 1`] = `
<figure
  className="base dark"
  id="about"
>
  <picture
    className="base picture dark"
  >
    <source
      media="(min-width: 1025px)"
      srcSet="https://release.aesop-web-ui.aesopdigital.io/images/products/Skin_Perfect_Facial_Hydrating_Cream_60mL_large.png"
    />
    <img
      alt="Image Alt Text"
    />
  </picture>
  <figcaption
    className="captionWrapper"
  >
    <div
      className="caption"
    >
      About Aesop
    </div>
  </figcaption>
</figure>
`;
